037-prototypes/proposal.md

```diff
@@ -0,0 +1,544 @@
+  way: [concourse/rfcs#9](https://github.com/concourse/rfcs/issues/9)
+
+
+## Glossary
+
+* **Prototype**: an implementation of the interface defined by this proposal,
+  typically packaged as an OCI container image.
+
+  Examples:
+
+  * a `git` prototype for interacting with commits and branches in a repo.
+  * a `time` prototype for performing interval triggers.
+
+* **Message**: an operation to perform against the object. Corresponds to a
+  command to run in the container image.
+
+  Examples:
+
+  * `check`
+  * `get`
+
+* **Object**: a logical entity, encoded as a JSON object, acted on and emitted
+  by a prototype in response to messages.
+
+  Examples:
+
+  * `{"uri":"https://github.com/concourse/rfcs"}`
+  * `{"ref":"e4be0b367d7bd34580f4842dd09e7b59b6097b25"}`
+
+* **Metadata**: additional information about an object to surface to users.
+
+  Examples:
+
+  * `[{"name":"committer","value":"Alex Suraci"}]`
+
+* **Bits**: a directory containing arbitrary data.
+
+  Examples:
+  
+  * source code checked out from a repo
+  * compiled artifacts created by a task or another prototype
+
+* **Resource**: a **prototype** which implements the following messages:
+
+  * `check`: discover objects representing versions, in order
+  * `get`: fetch an object version
+
+
+## Interface Types
+
+```go
+// Object is an object receiving messages and being emitted in response to
+// messages.
+type Object map[string]interface{}
+
+// InfoRequest is the payload written to stdin for the `./info` script.
+type InfoRequest struct {
+  // The properties specifying the object to act on.
+  Object Object `json:"object"`
+
+  // Configuration for handling TLS.
+  TLS TLSConfig `json:"tls,omitempty"`
+}
+
+// InfoResponse is the payload written to stdout from the `./info` script.
+type InfoResponse struct {
+  // The version of the prototype interface that this prototype conforms to.
+  InterfaceVersion string `json:"interface_version"`
+
+  // An optional icon to show to the user.
+  //
+  // Icons must be namespaced by in order to explicitly reference an icon set
+  // supported by Concourse, e.g. 'mdi:' for Material Design Icons.
+  Icon string `json:"icon,omitempty"`
+
+  // The messages supported by the object.
+  Messages []string `json:"messages,omitempty"`
+}
+
+// MessageRequest is the payload written to stdin for a message.
+type MessageRequest struct {
+  // The properties specifying the object to act on.
+  Object Object `json:"object"`
+
+  // Configuration for handling TLS.
+  TLS TLSConfig `json:"tls,omitempty"`
+
+  // Path to a file into which the message handler must write its response.
+  ResponsePath string `json:"response_path"`
+}
+
+// MessageResponse is written to the `response_path` for each object returned
+// by the message. Multiple responses may be written to the same file,
+// concatenated as a JSON stream.
+type MessageResponse struct {
+  // The identifying properties for the object.
+  Object Object `json:"object"`
```

Comment:
I wonder if there's room for a typed "error" in the `MessageResponse` 🤔 

E.g., if a `git` prototype receives a check message that asks for versions of a `svn` "repository", it could report back that error in the form of a typed "error" that could be gracefully interpreted 🤔 

or, at least a base "error prototype" that can be identified as an error but augmented however one sees fit

Reply:
:thinking: I agree in principle but I'm having a hard time thinking of cases where the error message would be semantically interpreted and 'handled' in some way. For example I don't really see a practical difference between returning an error message saying "your repo is svn but i'm a git resource" vs. just printing that to `stderr` and exiting nonzero.

Maybe validation errors could be returned as a data structure, but even those could just be printed to `stderr`, and it's kind of hard to specify a data structure here that will capture nuances across all prototypes.

I guess at this point I'm leaning towards under-specifying here and revising the proposal later (it's versioned!) if/when we have more information.

Reply:
Would revealing the error as an accessible field allow better control flow in your plan? Such as using the "on_error" hook for a specific type of error from InfoResponse.

Reply:
@kcmannem It might, but no one has asked for that feature (to my knowledge). It's hard to design for this kind of thing without having a good grasp of a use case that is also common enough to encode at this level. Personally that amount of control flow/error handling in a pipeline .yml kind of scares me. :sweat_smile: